spynnaker/pyNN/extra_algorithms/splitter_components/splitter_abstract_pop_vertex_neurons_synapses.py

```diff
@@ -0,0 +1,733 @@
+        sdram = MultiRegionSDRAM()
+        sdram.add_cost(
+            PopulationSynapsesMachineVertexLead.SYNAPSE_REGIONS.synapse_params,
+            max(app_vertex.get_synapse_params_size(), BYTES_PER_WORD))
+        sdram.add_cost(
+            PopulationSynapsesMachineVertexLead.SYNAPSE_REGIONS
+            .bitfield_builder,
+            max(exact_sdram_for_bit_field_builder_region(), BYTES_PER_WORD))
+        return sdram
+
+    def __proj_dependent_synapse_sdram(self, incoming_projections):
+        """ Get the SDRAM used by synapse cores dependent on the projections
+
+        :param list(~spynnaker.pyNN.models.Projection) incoming_projections:
+            The projections to consider in the calculations
+        :rtype: ~pacman.model.resources.MultiRegionSDRAM
+        """
+        app_vertex = self._governed_app_vertex
+        sdram = MultiRegionSDRAM()
+        sdram.add_cost(
+            PopulationSynapsesMachineVertexLead.SYNAPSE_REGIONS.pop_table,
+            max(MasterPopTableAsBinarySearch.get_master_population_table_size(
+                    incoming_projections), BYTES_PER_WORD))
+        sdram.add_cost(
+            PopulationSynapsesMachineVertexLead.SYNAPSE_REGIONS
+            .connection_builder,
+            max(app_vertex.get_synapse_expander_size(incoming_projections),
+                BYTES_PER_WORD))
+        sdram.add_cost(
+            PopulationSynapsesMachineVertexLead.SYNAPSE_REGIONS
+            .bitfield_filter,
+            max(get_estimated_sdram_for_bit_field_region(incoming_projections),
+                BYTES_PER_WORD))
+        sdram.add_cost(
+            PopulationSynapsesMachineVertexLead.SYNAPSE_REGIONS
+            .bitfield_key_map,
+            max(get_estimated_sdram_for_key_region(incoming_projections),
+                BYTES_PER_WORD))
+        return sdram
+
+    @property
+    def __get_max_delay(self):
+        if self.__max_delay is not None:
+            return self.__max_delay
+
+        # Find the maximum delay from incoming synapses
+        app_vertex = self._governed_app_vertex
+        max_delay_ms = 0
+        for proj in app_vertex.incoming_projections:
+            s_info = proj._synapse_information
+            proj_max_delay = s_info.synapse_dynamics.get_delay_maximum(
+                s_info.connector, s_info)
+            max_delay_ms = max(max_delay_ms, proj_max_delay)
+        max_delay_steps = math.ceil(max_delay_ms / machine_time_step_ms())
+        max_delay_bits = get_n_bits(max_delay_steps)
+
+        # Find the maximum possible delay
+        n_atom_bits = get_n_bits(app_vertex.get_max_atoms_per_core())
+        n_synapse_bits = get_n_bits(
+            app_vertex.neuron_impl.get_n_synapse_types())
+        n_delay_bits = MAX_RING_BUFFER_BITS - (n_atom_bits + n_synapse_bits)
+
+        # Pick the smallest between the two, so that not too many bits are used
+        final_n_delay_bits = min(n_delay_bits, max_delay_bits)
```

Comment:
This will only be used when the user hasn't specified the max delay.  If they do, it has to work or they get an error; if they don't it is automatically calculated here so they probably don't care.  The min part here decides the maximum delay possible on the core, but doesn't enable or disable delay extensions.  That is done in the following lines, again only if the user hasn't specified this.  

The rules currently are:
1. If the user doesn't specify max delay or whether to allow delay extensions, things are done automatically.  If things still don't work out (probably because even with a delay extension it can't be done - see delay adder), an error is generated.  If delays can be done without delay extensions according to the calculations here, delay extensions are disabled.  This appears to work in the tests done so far, so it seems to line up with delay adder calculations too.
1. If the user specifies a max delay, but doesn't specify whether to allow delay extensions, delay extensions are enabled.  This could then generate an error as above, or additionally if the neurons per core and max delay together can't be realised because of a lack of bits.
1. If the user doesn't specify max delay, but disallows delay extensions, this will generate an error if delays specified can't be realised without a delay extension (see delay adder).
1. If the user specifies both a max delay and whether to allow delay extensions, this is much the same as above cases in that errors might be generated here for a lack of bits or during delay adder for a lack of delay slots.

Reply:
Thanks - that makes sense.